File: dsa-js/arrays/two-sum.js
/**
 * <h2>Two Sum - Implementations</h2>
 * @see {@link https://neetcode.io/solutions/two-sum}
 */

/**
 * <h3>Approach 1: Hash Map (Optimal)</h3>
 * <ul>
 * <li><b>Time Complexity:</b> O(n)</li>
 * <li><b>Space Complexity:</b> O(n)</li>
 * </ul>
 */
function twoSumMap(nums, target) {
    const map = new Map();
    for (let i = 0; i < nums.length; i++) {
        const complementary = target - nums[i];
        if (map.has(complementary)) return [map.get(complementary), i];
        map.set(nums[i], i);
    }
    return [];
}

/**
 * <h3>Approach 2: Brute Force</h3>
 * <p>Checks every possible pair using nested loops.</p>
 * <ul>
 * <li><b>Time Complexity:</b> O(n²)</li>
 * <li><b>Space Complexity:</b> O(1)</li>
 * </ul>
 */
function twoSumBruteForce(nums, target) {
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) return [i, j];
        }
    }
    return [];
}

module.exports = {
    twoSumMap,
    twoSumBruteForce
};

File: dsa-python/arrays/two_sum.py
from typing import List


def two_sum_brute(nums: List[int], target: int) -> List[int]:
    """
    Brute Force Approach
    Time: O(n^2) | Space: O(1)
    """
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []

def two_sum_map(nums: List[int], target: int) -> List[int]:
    """
    Hash Map Approach (Optimal)
    Time: O(n) | Space: O(n)
    """
    seen = {}
    for i,n in enumerate(nums):
        complementary = target - n
        if complementary in seen:
            return [seen[complementary], i]
        seen[n] = i
    return []
